clippy_lints/src/functions/ref_option.rs

```diff
@@ -0,0 +1,108 @@
+use crate::functions::hir::GenericArg;
+use crate::functions::REF_OPTION;
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::source::snippet;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::intravisit::FnKind;
+use rustc_hir::{FnDecl, GenericArgs, ImplItem, MutTy, QPath, TraitItem, Ty, TyKind};
+use rustc_lint::LateContext;
+use rustc_span::{sym, Span};
+
+fn check_ty(cx: &LateContext<'_>, param: &Ty<'_>, fixes: &mut Vec<(Span, String)>) {
+    // TODO: is this the right way to check if ty is an Option?
+    if let TyKind::Ref(lifetime, MutTy { ty, .. }) = param.kind
+        && let TyKind::Path(QPath::Resolved(_, path)) = ty.kind
+        && path.segments.len() == 1
+        && let seg = &path.segments[0]
+        && seg.ident.name == sym::Option
+        // check if option contains a regular type, not a reference
+        // TODO: Should this instead just check that opt_ty is a TyKind::Path?
+        && let Some(GenericArgs { args: [GenericArg::Type(opt_ty)], .. }) = seg.args
+        && !matches!(opt_ty.kind, TyKind::Ref(..))
```

Comment:
Why wouldn't you want to lint for refs? This still constitutes a double reference, and an `Option<&T>` is the same size as a `&Option<&T>`, just with one more level of indirection. `TyKind::Never` might also be interesting here, but that's just a zero-sized type, so I doubt having a reference to that makes a lot of sense. The other options I don't see a reason to avoid linting them.